src/wp-includes/class-wp-modules.php

```diff
@@ -0,0 +1,246 @@
+ * @package WordPress
+ * @subpackage Modules
+ */
+
+/**
+ * Core class used to register modules.
+ *
+ * @since 6.5.0
+ */
+class WP_Modules {
+	/**
+	 * Holds the registered modules, keyed by module identifier.
+	 *
+	 * @since 6.5.0
+	 * @var array
+	 */
+	private $registered = array();
+
+	/**
+	 * Holds the module identifiers that were enqueued before registered.
+	 *
+	 * @since 6.5.0
+	 * @var array
+	 */
+	private $enqueued_before_registered = array();
+
+	/**
+	 * Registers the module if no module with that module identifier has already
+	 * been registered.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string            $module_identifier The identifier of the module. Should be unique. It will be used in the final import map.
+	 * @param string            $src               Full URL of the module, or path of the script relative to the WordPress root directory.
+	 * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this module. The dependencies can be strings or arrays. If they are arrays, they need an `id` key with the module identifier, and can contain a `type` key with either `static` or `dynamic`. By default, dependencies that don't contain a type are considered static.
+	 * @param string|false|null $version           Optional. String specifying module version number. Defaults to false. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, the version is the current timestamp. If $version is set to false, the version number is the currently installed WordPress version. If $version is set to null, no version is added.
+	 */
+	public function register( $module_identifier, $src, $dependencies = array(), $version = false ) {
+		if ( ! isset( $this->registered[ $module_identifier ] ) ) {
+			$deps = array();
+			foreach ( $dependencies as $dependency ) {
+				if ( isset( $dependency['id'] ) ) {
+					$deps[] = array(
+						'id'   => $dependency['id'],
+						'type' => isset( $dependency['type'] ) && 'dynamic' === $dependency['type'] ? 'dynamic' : 'static',
+					);
+				} elseif ( is_string( $dependency ) ) {
+					$deps[] = array(
+						'id'   => $dependency,
+						'type' => 'static',
+					);
+				}
+			}
+
+			$this->registered[ $module_identifier ] = array(
+				'src'          => $src,
+				'version'      => $version,
+				'enqueued'     => in_array( $module_identifier, $this->enqueued_before_registered, true ),
+				'dependencies' => $deps,
+			);
+		}
+	}
+
+	/**
+	 * Marks the module to be enqueued in the page.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $module_identifier The identifier of the module.
+	 */
+	public function enqueue( $module_identifier ) {
+		if ( isset( $this->registered[ $module_identifier ] ) ) {
+			$this->registered[ $module_identifier ]['enqueued'] = true;
+		} elseif ( ! in_array( $module_identifier, $this->enqueued_before_registered, true ) ) {
+			$this->enqueued_before_registered[] = $module_identifier;
```

Comment:
I would recommend to make the `$enqueued_before_registered` property a map of `$module_identifier => true`. The value wouldn't carry any meaning, but it's simply a small performance optimization, as it would allow us to use `isset()` instead of `in_array()` / `array_search()`, which is faster. Dequeuing would be faster too, as it could use `unset()` instead of `array_splice()`.

For instance here this would become:
```suggestion
		} elseif ( ! isset( $this->enqueued_before_registered[ $module_identifier ] ) ) {
			$this->enqueued_before_registered[ $module_identifier ] = true;
```

You could even avoid the `elseif` since unconditionally setting it to `true` would have the same outcome (as it's now an associative array where the same module couldn't appear multiple times).
```suggestion
		} else {
			$this->enqueued_before_registered[ $module_identifier ] = true;
```

Reply:
Makes sense, thank you 🙂

Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/66987f0f5df80620e1f6982313e6916b322f941c.